constantine/math_compiler/pub_fields.nim

```diff
@@ -24,41 +210,879 @@ proc genFpAdd*(asy: Assembler_LLVM, fd: FieldDescriptor): string =
+  ## with `r` the resulting negated field element.
+  ## The negation is only performed if `condition` is `true`.
+  ##
+  ## Generates a call, so that we one can use this proc as part of another procedure.
+  let name = fd.name & "_cneg_internal"
+  asy.llvmInternalFnDef(
+          name, SectionName,
+          asy.void_t, toTypes([r, a, c]),
+          {kHot}):
+
+    tagParameter(1, "sret")
+    let (ri, ai, ci) = llvmParams
+    let M = asy.getModulusPtr(fd)
+
+    # first call the regular negation
+    asy.neg_internal(fd, ri, ai)
+    # now ccopy
+    asy.ccopy_internal(fd, ri, ai, asy.br.`not`(ci))
+
+    asy.br.retVoid()
+
+  asy.callFn(name, [r, a, c])
+
+proc genFpCNeg*(asy: Assembler_LLVM, fd: FieldDescriptor): string =
+  ## Generate an internal conditional out of place field negation
+  ## with signature
+  ##   `void name(FieldType r, FieldType a, bool condition)`
+  ## with `r` the resulting negated field element.
+  ## The negation is only performed if `condition` is `true`.
+  ##
+  ## Returns the corresponding name to call it
+  let name = fd.name & "_cneg"
+  asy.llvmPublicFnDef(name, "ctt." & fd.name, asy.void_t, [fd.fieldTy, fd.fieldTy, asy.ctx.int1_t()]):
+    let (r, a, c) = llvmParams
+    asy.cneg_internal(fd, r, a, c)
+    asy.br.retVoid()
+  return name
+
+template genCountdownLoop(asy: Assembler_LLVM, fn, initialValue: ValueRef, body: untyped): untyped =
+  ## A helper template that constructs a `countdown(initialValue, 1)` loop.
+  let loopEntry = asy.ctx.appendBasicBlock(fn, "loop.entry")
+  let loopBody  = asy.ctx.appendBasicBlock(fn, "loop.body")
+  let loopExit  = asy.ctx.appendBasicBlock(fn, "loop.exit")
+
+  # Branch to loop entry
+  asy.br.br(loopEntry)
+
+  let c0 = constInt(asy.ctx.int32_t(), 0)
+  let c1 = constInt(asy.ctx.int32_t(), 1)
+
+  # Loop entry
+  asy.br.positionAtEnd(loopEntry)
+  let condition = asy.br.icmp(kSGT, initialValue, c0)
+  asy.br.condBr(condition, loopBody, loopExit)
+
+  # Loop body
+  asy.br.positionAtEnd(loopBody)
+  let phi = asy.br.phi(getTypeOf initialValue)
+  phi.addIncoming(initialValue, loopEntry)
+
+  # Decrement
+  let decremented = asy.br.sub(phi, c1)
+  phi.addIncoming(decremented, loopBody)
+
+  # The loop body
+  body
+
+  # Check if we should continue looping
+  let continueLoop = asy.br.icmp(kSGT, decremented, c0)
+  asy.br.condBr(continueLoop, loopBody, loopExit)
+
+  # Loop exit
+  asy.br.positionAtEnd(loopExit)
+
+proc genFpNsqrRT*(asy: Assembler_LLVM, fd: FieldDescriptor): string =
+  ## Generate a public field n-square proc
+  ## with signature
+  ##   `void name(FieldType r, FieldType a, count int)`
+  ## with `r` the resulting field element, `a` the field element to square
+  ## `count` times (i.e. for counts `[1, 2, 3, ...]` it yields `[a², a⁴, a⁸, ...]`).
+  ##
+  ## Returns the corresponding name to call it
+  let name = fd.name & "_nsqr"
+  let ptrBool = pointer_t(asy.ctx.int1_t())
+  ## XXX: int32? int64?
```

Comment:
`int32`, no one does 4 billions iterations of field squaring. RSA goes to 4096.
And it's way way faster on GPUs to use 32-bit numbers.